### PR TITLE
Emit balance change events from the protocol

### LINKED ISSUE #84: protocol balance change events
The ledger should attach an event to blocks in which some accounts' balances are modified by the protocol (e.g. wrapper tx fee charges, PoS slashing, future governance proposal refunds), to make the changes externally visible and index-able.

### PR DESCRIPTION
**Pull request by sug0**
## Describe your changes

Closes #84

Emits balance change events for the following protocol actions:

- Fee payments to validators
- PGF payments (including over IBC)
- Governance refunds
- IBC token minting/burning
- Proof-of-stake validator slashing
- WASM transfers

## Indicate on which release or other PRs this topic is based on

#3102

Diff for review: https://github.com/anoma/namada/compare/tiago/move-events-to-submodules..tiago/balance-change-events (https://github.com/anoma/namada/pull/3141/files/68420a57f251ca8e68ae6960a4a0081e57c16403..406f715b416b069b310410fec20f0ba0cb7c4a06)

## Checklist before merging to `draft`
- [x] I have added a changelog
- [x] Git history is in acceptable state

### CODE REVIEW

File: crates/state/src/wl_state.rs [hunk 1]
```diff
@@ -1183,6 +1232,30 @@ where
     }
 }
 
+impl<D, H> EmitEvents for TempWlState<'_, D, H>
```
Comment (grarco) [inline crates/state/src/wl_state.rs:1235]: The `TempWlState` is only used for validation/simulation in `mempool_validate` and `prepare/process_proposal`. Do we need to support events in those cases too?

File: crates/trans_token/src/event.rs [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+
+    /// Transfer token event.
+    pub const TRANSFER: EventType = event_type!(TokenEvent, "transfer");
+}
+
+/// A user account.
+#[derive(Debug, Clone, Ord, PartialOrd, Eq, PartialEq, Hash)]
+pub enum UserAccount {
+    /// Internal chain address in Namada.
+    Internal(Address),
```
Comment (grarco) [inline crates/trans_token/src/event.rs:33]: Can this cause some confusion with the enum variant we have on the `Address` type?

**Review by grarco — DISMISSED** [review]
A couple of minor comments. LGMT!

File: crates/state/src/wl_state.rs (hunk 1 above)
Comment (sug0) [reply crates/state/src/wl_state.rs:1235]: ah, I guess not. I implemented `EmitEvents` on pretty much every type that also implemented `State`. I'll remove this impl

File: crates/trans_token/src/event.rs (hunk 2 above)
Comment (sug0) [reply crates/trans_token/src/event.rs:33]: hmm.. any other name suggestions?

Comment (grarco) [reply crates/trans_token/src/event.rs:33]: I'm terrible at naming. Let's keep it like that 👍🏻

**Comment by brentstone** [comment]
FWIW, currently slashing does not actually involve any balance changes. The tokens are held in the PoS account and they are kept there, only other voting power-related data is manipulated. In PoS, only bonding and withdrawing involve balance changes.

I don't think I know enough about the event emission in general to say whether or not we should still emit events for slashing or if they should strictly be done for balance changes in the context of this PR though.

**Comment by sug0** [comment]
@brentstone yeah, slashing events were loosely grouped together with other balance change events in this pr, but they're defined as Proof-of-Stake events, rather than Token events. I think it's still useful to emit them in the protocol, as long as we distinguish them from balance change events.

**Comment by Fraccaman** [comment]
my only comment is that we could move each `TokenTransfer` event to a static method to avoid having constant string around the code, but we can also do this refactor later. Anything else looks good to me!

**Review by Fraccaman — APPROVED** [review]

File: crates/trans_token/Cargo.toml [hunk 3]
```diff
@@ -20,8 +20,10 @@ migrations = [
 
 [dependencies]
 namada_core = { path = "../core" }
+namada_events = { path = "../events", default-features = false }
```
Comment (tzemanovic) [inline crates/trans_token/Cargo.toml:23]:
```suggestion
namada_events = { path = "../events" }
```
tiny (there seem to be no default features)

File: crates/token/src/lib.rs [hunk 4]
```diff
@@ -9,6 +9,10 @@ pub mod storage_key {
     pub use namada_trans_token::storage_key::*;
 }
 
+pub mod event {
+    pub use namada_trans_token::event::*;
+}
```
Comment (tzemanovic) [inline crates/token/src/lib.rs:14]:
```suggestion
pub use namada_trans_token::event;
```

**Review by tzemanovic — DISMISSED** [review]

File: crates/trans_token/Cargo.toml (hunk 3 above)
Comment (sug0) [reply crates/trans_token/Cargo.toml:23]: I added this because eventually I'd like to move the event log out of `namada_sdk` to this crate

**Pull request merged**